### DISCUSSION TITLE
Are GraalVM Native Images supported?

### DISCUSSION THREAD

**Discussion by konstantinblaesi** [Netflix/dgs-framework#1264] [General]
Can the application be built as a GraalVM Native Image? I know that Spring Boot started working on it and probably has more or less support for it. How is it when adding DGS into the mix?

**Comment by paulbakker** [comment]
It's not something we've looked into at all. I would be surprised if DGS
specifically has any major issues that would make this more challenging,
but we haven't tested it.

**Reply by To-da** [reply]
DGS do not work out of the box with Spring native. Example of exception when single query is defined:
`Caused by: com.netflix.graphql.dgs.exceptions.InvalidDgsConfigurationException: Parent type Query not found, but it was referenced on com.netflix.graphql.dgs.internal.DgsSchemaProvider in @DgsData annotation for field accounts`

where query is implemented in `@DgsComponent` as:
```kotlin
    @DgsQuery
    fun accounts(env: DataFetchingEnvironment): Connection<Foo> =
        SimpleListConnection(fooService.foos()).get(env)
```

Not sure what `@TypeHint` to use to make it work. More investigations needs to be done by Tracing Agent.

**Reply by Wallman** [reply]
More examples:
> The program tried to reflectively invoke method public final com.netflix.graphql.dgs.webmvc.autoconfigure.DgsWebMvcConfigurationProperties$DgsSchemaJsonConfigurationProperties com.netflix.graphql.dgs.webmvc.autoconfigure.DgsWebMvcConfigurationProperties.getSchemaJson() without it being registered for runtime reflection. Add it to the reflection metadata to solve this problem

> Property or field 'path' cannot be found on object of type 'com.netflix.graphql.dgs.webmvc.autoconfigure.DgsWebMvcConfigurationProperties$DgsSchemaJsonConfigurationProperties

> The program tried to reflectively invoke method public graphql.schema.idl.RuntimeWiring$Builder sh.vcm.graphql.dgs.ScalarsRegistrator.addScalar(graphql.schema.idl.RuntimeWiring$Builder) without it being registered for runtime reflection.

Running Spring Boot 3.2.2 and DGS 8.24

**Reply by asavov** [reply]
More examples:
Spring AOT requires compilation with **-parameters**
```
org.springframework.boot.context.properties.bind.MissingParametersCompilerArgumentException:
     Constructor binding in a native image requires compilation with -parameters
     but the following classes were compiled without it:
        com.netflix.graphql.dgs.metrics.micrometer.DgsGraphQLMetricsProperties$CardinalityLimiterKind
```

**Workaround**: exclude DGS micrometer auto-config 
`@SpringBootApplication(exclude = DgsGraphQLMicrometerAutoConfiguration.class)`

**Env**: SB 3.1.8, DGS 8.2.2

**Ask**: may the DGS code be compiled with **-parameters** ?

**Comment by federico-s** [comment]
Hello, any update on this with Spring Boot 3.0.4 and DGS 6.0.1?

**Comment by srinivasankavitha** [comment]
Hi - we haven't tested this and it's not on our roadmap for the near
future, since this is not a priority for us internally. We do accept
contributions, so would welcome any PRs needed to enable this functionality.

**Comment by MrOstling** [comment]
I can confirm that DGS doesn't work out of the box with graalvm native images. We were able to get it to work for us by adding the native-image-agent to the test runner configuration in the build.gradle.kts 
```
   tasks {
        test {
            useJUnitPlatform()
            jvmArgs("-agentlib:native-image-agent=config-output-dir=./src/main/resources/META-INF/native-image")
        }
    }
 ```
 Once this line was there we ran the tests with a native-image compatible java version (graalvm) using `./gradlew clean build` this generated all the hints we needed for native compile. We then ran `./gradlew assemble` to get those generated files into the jar. Once that was done it is just a matter of publishing.

**Comment by Wallman** [comment]
What is the status on GraalVM/Native for DGS nowadays? On the roadmap? Still not prioritized?

**Reply by jaizquierdogalan** [reply]
By gaining a better understanding of the state of GraalVM, itcan unlock significant resource savings in production. Don't miss out on this opportunity to optimize the operations and improve the bottom line.

**Reply by srinivasankavitha** [reply]
We don't have this on the roadmap yet due to higher priorities. Thanks for the input and votes so far, we will post an update once we have a better idea of timelines and support for this.
